pkg/limits/frontend/ring.go

```diff
@@ -56,50 +66,111 @@ func (g *ringGatherer) exceedsLimits(ctx context.Context, req *proto.ExceedsLimi
 	if err != nil {
 		return nil, err
 	}
-	partitionConsumers, err := g.getPartitionConsumers(ctx, rs.Instances)
+	// Get the partition consumers for each zone.
+	zonesPartitions, err := g.getZoneAwarePartitionConsumers(ctx, rs.Instances)
 	if err != nil {
 		return nil, err
 	}
-	ownedStreams := make(map[string][]*proto.StreamMetadata)
-	for _, s := range req.Streams {
-		partition := int32(s.StreamHash % uint64(g.numPartitions))
-		addr, ok := partitionConsumers[partition]
+	// In practice we want zones to be queried in random order to spread
+	// reads. However, in tests we want a deterministic order so test cases
+	// are stable and reproducible. Having a custom sort func supports both
+	// use cases as zoneCmp can be switched out in tests.
+	zonesToQuery := make([]string, 0, len(zonesPartitions))
+	for zone := range zonesPartitions {
+		zonesToQuery = append(zonesToQuery, zone)
+	}
+	slices.SortFunc(zonesToQuery, g.zoneCmp)
+	// Make a copy of the streams from the request. We will prune this slice
+	// each time we receive the responses from a zone.
+	streams := make([]*proto.StreamMetadata, 0, len(req.Streams))
+	for _, stream := range req.Streams {
+		streams = append(streams, stream)
+	}
+	// Query each zone as ordered in zonesToQuery. If a zone answers all
+	// streams, the request is satisifed and there is no need to query
+	// subsequent zones. If a zone answers just a subset of streams
+	// (i.e. the instance that is consuming a partition is unavailable or the
+	// partition that owns one or more streams does not have a consumer)
+	// then query the next zone for the remaining streams. We repeat this
+	// process until all streams have been queried or we have exhausted all
+	// zones.
+	responses := make([]*proto.ExceedsLimitsResponse, 0)
+	for _, zone := range zonesToQuery {
+		resps, answered, err := g.doExceedsLimitsRPCs(ctx, req.Tenant, streams, zonesPartitions[zone])
+		if err != nil {
+			continue
+		}
+		responses = append(responses, resps...)
+		// Remove the answered streams from the slice. The slice of answered
+		// streams must be sorted so we can use sort.Search to subtract the
+		// two slices.
+		slices.Sort(answered)
+		streams = slices.DeleteFunc(streams, func(stream *proto.StreamMetadata) bool {
+			// see https://pkg.go.dev/sort#Search
+			i := sort.Search(len(answered), func(i int) bool {
+				return answered[i] >= stream.StreamHash
+			})
+			return i < len(answered) && answered[i] == stream.StreamHash
+		})
+		// All streams been checked against per-tenant limits.
+		if len(streams) == 0 {
+			break
+		}
+	}
+	// TODO(grobinson): In a subsequent change, I will figure out what to do
```

Comment:
At present, unanswered streams are permitted. First I want to emit a metric to track this, and I will do this in a follow up PR to make the change easier to see.